src/odemis/odemisd/main.py

```diff
@@ -90,40 +92,54 @@ def __init__(self, model_file, settings_file, create_sub_containers=False,
             logging.exception("When instantiating file %s", self._model.name)
             raise IOError("Unexpected error at instantiation")
 
-    def _on_va_change(self, _):
+    def _observe_persistent_va(self, comp, prop_name):
         """
-        Callback to VA change, calls write_persistent_data
+        Listen to changes in persistent va, update the value of instantiator.persistent_props
+        and write data to file.
+        comp (HwComponent): component
+        prop_name (str): name of va
         """
-        self._write_persistent_data()
+        def on_va_change(value, comp=comp.name, prop=prop_name):
+            # Update persistent properties in instantiator
+            if comp not in self._instantiator.persistent_props:
+                self._instantiator.persistent_props[comp] = {}
+            self._instantiator.persistent_props[comp][prop] = value
+            self._write_persistent_data()
+
+        va = getattr(comp, prop_name)
+        va.subscribe(on_va_change)
+        self.persistent_listeners.append(on_va_change)
 
     def _write_persistent_data(self):
         """
         Write values for all persistent properties and metadata to the settings file.
         """
         data = {}
-        for comp in self._instantiator.components:
-            props, mds = self._instantiator.get_persistent(comp)
-            # Initialize internal dict
-            comp_data = {}
-            if props:
-                comp_data['properties'] = {}
-            if mds:
-                comp_data['metadata'] = {}
-
-            for prop in props:
-                comp_data['properties'][prop] = getattr(comp, prop).value
-            for md in mds:
-                fullname = "MD_" + md
-                comp_data['metadata'][md] = comp.getMetadata()[getattr(model, fullname)]
-
-            # Only store components that actually contain persistent data
-            if comp_data:
-                data[comp.name] = comp_data
+        for comp, prop_dict in self._instantiator.persistent_props.items():
+            data[comp] = {}
+            data[comp]['properties'] = prop_dict
+        for comp, md_dict in self._instantiator.persistent_mds.items():
+            if comp not in data:
+                data[comp] = {}
+            data[comp]['metadata'] = md_dict
 
         self._settings.truncate(0)  # delete previous file contents
         self._settings.seek(0)  # go back to position 0
         yaml.safe_dump(data, self._settings)
 
+    def _update_persistent_metadata(self):
+        """
+        Update all elements of instantiator.persistent_mds
+        """
+        for comp in self._instantiator.components:
+            _, mds = self._instantiator.get_persistent(comp.name)
+            for md in mds:
+                fullname = "MD_" + md
+                try:
+                    self._instantiator.persistent_mds[comp.name][md] = comp.getMetadata()[getattr(model, fullname)]
+                except Exception as ex:
```

Comment:
You could catch the KeyError exception specifically, as it's probably going to be the most common error, to show a clearer logging message like: `Persistent metadata %s not found on component %s`.